# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsIGlobalObject.h
# Commit: a8e623b228ae
# Full Hash: a8e623b228aeb04c741bebeeaea4f2524b2ddcc7
# Author: Tom Ritter <tom@mozilla.com>
# Date: 2022-11-29 21:47:49
# Regressor Bug: 1778510
# File Overlap Count: 1
# Description:
#   Bug 1778510: Add RTPCallerType to GlobalObject r=asuth
#   
#   This centralizes the logic in one place.
#   
#   In order to do this, we will need to check the principal
# ==============================================================================

diff -r 32fdf6f014f7 -r a8e623b228ae dom/base/nsIGlobalObject.h
--- a/dom/base/nsIGlobalObject.h	Tue Nov 29 13:34:14 2022 +0000
+++ b/dom/base/nsIGlobalObject.h	Tue Nov 29 13:34:15 2022 +0000
@@ -13,6 +13,7 @@
 #include "mozilla/dom/DispatcherTrait.h"
 #include "mozilla/dom/ServiceWorkerDescriptor.h"
 #include "mozilla/OriginTrials.h"
+#include "nsContentUtils.h"
 #include "nsHashKeys.h"
 #include "nsISupports.h"
 #include "nsStringFwd.h"
@@ -58,6 +59,17 @@
 class ModuleLoaderBase;
 }  // namespace JS::loader
 
+
+// Reduce Timer Precision (RTP) Caller Type
+// This lives here because anything dealing with RTPCallerType determines it
+// through this object.
+enum class RTPCallerType : uint8_t {
+  Normal = 0,
+  SystemPrincipal = (1 << 0),
+  ResistFingerprinting = (1 << 1),
+  CrossOriginIsolated = (1 << 2)
+};
+
 /**
  * See <https://developer.mozilla.org/en-US/docs/Glossary/Global_object>.
  */
@@ -134,7 +146,7 @@
   bool HasJSGlobal() const { return GetGlobalJSObjectPreserveColor(); }
 
   // This method is not meant to be overridden.
-  nsIPrincipal* PrincipalOrNull();
+  nsIPrincipal* PrincipalOrNull() const;
 
   void RegisterHostObjectURI(const nsACString& aURI);
 
@@ -244,6 +256,8 @@
    */
   virtual bool ShouldResistFingerprinting() const = 0;
 
+  RTPCallerType GetRTPCallerType() const;
+
   /**
    * Threadsafe way to get nsIPrincipal::GetHashValue for the associated
    * principal.
@@ -268,6 +282,8 @@
  protected:
   virtual ~nsIGlobalObject();
 
+  virtual bool IsSystemPrincipal() const;
+
   void StartDying() { mIsDying = true; }
 
   void StartForbiddingScript() { mIsScriptForbidden = true; }